src/services/refactors/convertStringOrTemplateLiteral.ts

```diff
@@ -0,0 +1,189 @@
+/* @internal */
+namespace ts.refactor.convertStringOrTemplateLiteral {
+    const refactorName = "Convert string concatenation or template literal";
+    const toTemplateLiteralActionName = "Convert to template literal";
+    const toStringConcatenationActionName = "Convert to string concatenation";
+
+    const refactorDescription = getLocaleSpecificMessage(Diagnostics.Convert_string_concatenation_or_template_literal);
+    const toTemplateLiteralDescription = getLocaleSpecificMessage(Diagnostics.Convert_to_template_literal);
+    const toStringConcatenationDescription = getLocaleSpecificMessage(Diagnostics.Convert_to_string_concatenation);
+
+    registerRefactor(refactorName, { getEditsForAction, getAvailableActions });
+
+    function getAvailableActions(context: RefactorContext): ReadonlyArray<ApplicableRefactorInfo> {
+        const { file, startPosition } = context;
+        const node = getNodeOrParentOfBraces(file, startPosition);
+        const maybeBinary = getParentBinaryExpression(node);
+        const actions: RefactorActionInfo[] = [];
+
+        if ((isBinaryExpression(maybeBinary) || isStringLiteral(maybeBinary)) && isStringConcatenationValid(maybeBinary)) {
+            actions.push({ name: toTemplateLiteralActionName, description: toTemplateLiteralDescription });
+        }
+
+        if (isTemplateLike(node)) {
+            actions.push({ name: toStringConcatenationActionName, description: toStringConcatenationDescription });
+        }
+
+        return [{ name: refactorName, description: refactorDescription, actions }];
+    }
+
+    function getNodeOrParentOfBraces(file: SourceFile, startPosition: number) {

```

Comment:
it's required so that parentheses can be used for selection

``"Panther is " + (43) + " years old";``
* 43 and ( ) can be selected to trigger code action

``"Panther is " + (43 + 3) + " years old";``
* only ( ) can be selected to trigger code action

``"Panther is " + (43 + 3 + "e") + " years old";``
* () selection tranforms outer scope
* any selection of 43+3+"e" transforms only inside of parentheses